tests/ts/examples.ts

```diff
@@ -584,6 +607,132 @@ Person.query().where(ref('Model.jsonColumn:details'), '=', lit({ name: 'Jennifer
+  .findOne({ firstName: 'Mo' })
+  .whereJsonIsArray('x:y')
+  .throwIfNotFound();
+const patchWhere: Promise<number> = qb.patch({ firstName: 'Mo' }).where('id', 32);
+const patchWhereIn: Promise<number> = qb.patch({ firstName: 'Mo' }).whereIn('id', [1, 2, 3]);
+const patchWhereJson: Promise<number> = qb
+  .patch({ firstName: 'Mo' })
+  .whereJsonSupersetOf('x:y', 'abc');
+const patchWhereJsonIsArray: Promise<number> = qb
+  .patch({ firstName: 'Mo' })
+  .whereJsonIsArray('x:y');
+const patchThrow: Promise<number> = qb.patch({ firstName: 'Mo' }).throwIfNotFound();
+const updateWhere: Promise<number> = qb.update({ firstName: 'Mo' }).where('id', 32);
+const updateWhereIn: Promise<number> = qb.update({ firstName: 'Mo' }).whereIn('id', [1, 2, 3]);
+const updateWhereJson: Promise<number> = qb
+  .update({ firstName: 'Mo' })
+  .whereJsonSupersetOf('x:y', 'abc');
+const updateWhereJsonIsArray: Promise<number> = qb
+  .update({ firstName: 'Mo' })
+  .whereJsonIsArray('x:y');
+const updateThrow: Promise<number> = qb.update({ firstName: 'Mo' }).throwIfNotFound();
+const deleteWhere: Promise<number> = qb.delete().where('lastName', 'like', 'Mac%');
+const deleteWhereIn: Promise<number> = qb.delete().whereIn('id', [1, 2, 3]);
+const deleteThrow: Promise<number> = qb.delete().throwIfNotFound();
+const deleteByIDThrow: Promise<number> = qb.deleteById(32).throwIfNotFound();
+
+// The location of `first` doesn't matter.
+
+const whereFirst: Promise<Person | undefined> = qb.where({ firstName: 'Mo' }).first();
+const firstWhere: Promise<Person | undefined> = qb.first().where({ firstName: 'Mo' });
+
+// Returning restores the result to Model or Model[].
+
+const whereInsertRet: Promise<Person> = qb
+  .where({ lastName: 'MacMoo' })
+  .insert({ firstName: 'Mo' })
+  .returning('dbGeneratedColumn');
+const whereMultiInsertRet: Promise<Person[]> = qb
+  .where({ lastName: 'MacMoo' })
+  .insert([{ firstName: 'Mo' }, { firstName: 'Bob' }])
+  .returning('dbGeneratedColumn');
+const whereUpdateRet: Promise<Person[]> = qb
+  .where({ lastName: 'MacMoo' })
+  .update({ firstName: 'Bob' })
+  .returning('dbGeneratedColumn');
+const wherePatchRet: Promise<Person[]> = qb
+  .where({ lastName: 'MacMoo' })
+  .patch({ firstName: 'Mo' })
+  .returning('age');
+const whereDelRetFirstWhere: Promise<Person | undefined> = qb
+  .delete()
+  .returning('lastName')
+  .first()
+  .where({ firstName: 'Mo' });
+
+// Verify that Model.query() and model.$query() return the same type of query builder.
+// Confirming this prevent us from having to duplicate the tests for each.
+
+async function checkQueryEquivalence() {
+  // Confirm that every $query() type is a query() type
+
+  let staticQB = Person.query()
+    .first()
+    .throwIfNotFound();
+  const person = await staticQB;
+  staticQB = person.$query();
+
+  // Confirm that every query() type is a $query() type
+
+  let instanceQB = person.$query();
+  instanceQB = staticQB;
+}
+
```

Comment:
@mceachen, I think I managed to avoid code duplication entirely, as well as make no assumptions about implementation. What do you think of this?